### ISSUE TITLE
Should we merge 'contexts' within an FDC3 API Channel

### ISSUE THREAD

**Issue by lspiro-Tick42** [finos/FDC3#152]
### Description of Problem:
FDC3 v1.1 is introducing the idea of [Channels](https://github.com/finos/FDC3/blob/master/src/api/channel/channel.ts). Various Channel styles are supported but a common one is to allow users to assign application windows to a channel identified by a Colour. The Channel API allows an application to set the Context for a Channel and to get the Current Context for a Channel. The Context data type is not defined, but I believe it is expected to be an [FDC3 Context](https://github.com/finos/FDC3/tree/master/src/context/schemas) expressed as JSon, but this is not specifiied in the API.  

When there are multiple independent applications connected to the same channel, and the applications have different contexts do we need to specify how the different contexts are specified? And is each update an overwrite of the previous Context or should we try and merge Contexts?

For example say we have a number of independent applications that can 'set the Context' on a channel, and then various 'listeners' that try and display data related to the 'current context'.

Say one 'setter' is a CRM and it will set context using the [Contact Schema](https://github.com/finos/FDC3/blob/master/src/context/schemas/contact.schema). A second 'setter' is a Watchlist app and it sets context using the [Instrument Schema](https://github.com/finos/FDC3/blob/master/src/context/schemas/instrument.schema).

We have a number of listeners, say a chart and a news app that can synchronise on an  'Instrument Context' and a 'Call History' app that shows recent communication with a Contact.

If the user clicks on an Instrument in the Watch list, the Chart and News apps should synchronise on the Instrument and the Call History window is unchanged.

The issue is what should happen when the user then selects a Contact in the CRM (and the CRM sets the Channel context to be a Contact).

Is the 'current Instrument' context deleted ? Or should the Channel Context be able to hold both a Contact and an Instrument, which implies that when an application sets a Channel context using the broadcast method, this is an update. This also implies that getCurrentContext on a channel needs to return a 'list of contexts'

If it is decided that a Channel context can hold both an Instrument and a Contact then we also need to discuss how a composite context like Portfolio which could hold both Instrument and Contact should be handled. 

### Potential Solutions:
The Context for a Channel should be a dictionary of Context types, so that it can hold a single instance of Instrument, Contact etc. 

When an application uses 'composite' context like Portfolio it should set the Context for Portfolio, Instrument and Contact.

**Comment by donbasuno** [comment]
Hi @lspiro-Tick42,

It's an interesting concept but as per normal my take is that we should start with the basics. 

Channels will allow us to place applications that do handle the same context into a workflow and thus separate eg the "instrument workflow" and the "contacts workflow" and the channel simply holds the "latest" context. I am sure it will happen - but putting apps that handle Contacts (CRM, Call History) and Instruments (Watchlist, Chart and News) seems like a bit o an edge case right now? 

Happy to hear thoughts from others as well but if you do think this is key functionality now think a more detailed use case would be the best next step.

In any case - similar situations can happen even when the context is the same - eg app A sets the context to an Instrument containing only a SEDOL, App B later sets the context to an Instrument containing only an ISIN. App C get the context but can only handle SEDOL so will not be able to show data for the latest context. So in that sense you bring up an interesting point on what to do when an app has not been able to update based on the latest context.

Best regards,

Johan

**Comment by rikoe** [comment]
@lspiro-Tick42 my 2 cents is that this is far too much complexity to add to a standard.

The context listener will tell you when a particular context is updated (or what the current value is). There is every chance we will end up with the ability to differentiate context items by type when setting up listeners (see #121).

Beyond that, if a particular framework or app wants to do complicated things like merging contexts together, that is a possibility, but I don't see this supported by the standard out of the box.

**Comment by lspiro-Tick42** [comment]
I think the key thing is to be very clear in our documentation about expected usage as well as the strict description of the individual functions. 

Expected usage is important, especially where we expect applications from multiple vendors to "just work". 

If we do not want to support the case where an app in a channel may set a composite context which other apps in the channel may chose to read (and maybe set) only parts of the composite, then we should state that in the relevant 'usage note'

**Comment by rikoe** [comment]
@lspiro-Tick42 I am not sure I agree. If we include everything the standard _doesn't_ do, it will make for a very long document!

Anything we support should be explicit (and unambiguous) in the standard. If it isn't in there, it isn't supported (which doesn't restrict an implementer from extending the behaviour).

@nkolba what's your perspective?

**Comment by nkolba** [comment]
I agree with the sentiment that we should detail expected usage/use cases more.  

However, I agree 100% with @rikoe that we should NOT do this from the perspective of what the standard doesn't do.  

@lspiro-Tick42 - your use case above - while not formalized in the standard could be implemented with a custom context type and behavior on the part of your platform without violating anything in the spec.

**Comment by nkolba** [comment]
Following up on the discussion in the WG earlier today. 

Here is how the current 1.1 spec addresses the use cases raised.

- The spec does not preclude a desktop agent from treating the current context as a dictionary of different types.   In fact, the spec explicitly mentions this as an [implementation option](https://fdc3.finos.org/docs/next/api/ref/Channel): *Desktop agent implementations may decide to record most recent contexts by type, in which case it will be possible to get the most recent context of each type, but this is not necessarily guaranteed.* 
- An app can set multiple context listeners filtering on different types and query current context on a channel multiple times filtering on multiple context types. 

This could be implemented by the desktop agent as a single list that filters by context-type and returns the latest with the matching type (if specified) or as multiple entries keyed by context-type, and certainly in a number of other ways.  I don't think we want to tell platform builders how they should do this. 

If we wanted to add a method to return all current context-types for a channel in one shot, this certainly could be done without breaking backwards compatibility of anything.  Given that an app should know what types its interested in, this doesn't seem pressing.

@lspiro-Tick42 If there are blockers here for 1.1, can we please get those detailed on this thread?

**Comment by mattjamieson** [comment]
There is definitely value in supporting filtering of context subscriptions; practically speaking I'm fairly sure we all must end up doing this today within the context handler.

I think we should consider making more explicit the expected behaviour of `getCurrentContext()` when called with no type.

One option could be to separate the two features into separate functions:
```
// Returns the last context broadcast matching the supplied type.
// Returns null if no context has been broadcast for that type.
getCurrentContext(contextType: string): Promise<Context|null>;
```
```
// Returns the last context broadcast regardless of type.
// Returns null if no context has been broadcast on the channel.
getLastContext(): Promise<Context|null>
```

This would still leave the implementer to decide whether to store context keyed by type. A naive implementation could be:
```
getCurrentContext(contextType: string): Promise<Context> {
  return getLastContext()
    .then(c => c.type === contextType ? c : null);
}
```

This does raise the question of how long context types live, and how you clear a context of a particular type from a channel. The TTL is probably an implementation concern, but perhaps a `clear` function should be part of the API (though probably not a concern for 1.1).

```
clearContext(contextType: string): Promise<void>;
```

The discussion of context adaption and merging is an interesting one, but I think should be for after 1.1, and perhaps deserves an issue in its own right?

**Comment by lspiro-Tick42** [comment]
My preference is that a Context should be a dictionary of FDC3 Context values, indexed by type. So that a context can hold a current value for a set of context types, and I would make this mandatory for implementations. I understand that this is not supported by the WG, although I would imagine many of us are doing this in our implementations.

If we wish to keep supporting a value per context type an optional implementation detail, then I think Matt’s suggestion of creating a separate functions to access the last value set independent of type and to make the getCurrentContext typed is a good idea.

In which case can I suggest that we also add an additional method

// Return all the values held in the current context.
// Returns null if not context has been broadcast on the channel
// NB It is an implementation decisions as to whether a platform choses to allow a channel Context to store more than one value.
getContextValues() : Promise<Array<Context>|null>;

I also think that Clear is a useful method since there are certain UI actions where that is useful.

Leslie

**Comment by nkolba** [comment]
@mattjamieson 

I totally agree that getting and handling context by type will be the dominant pattern.

I am all for making the API as self-documenting as possible, what I'm not sure of, is what is a stronger naming then what we already have. 

The distinction between "Current" and "Last" seems unclear.   From the outside, one might guess that "getLast" would return the context that happened before the current one - rather than the current context regardless of type.  

A more precise naming might be  "getCurrentContext" and "getCurrentContextByType" - but this is quite verbose - using the overload seems a lot more DRY and is in keeping with addContextListener.  

Definitely implementors will need to think about TTL on current context - especially in a multi-context world.  The clearContext API is an interesting approach to the problem.  Definitely something to consider post 1.1.

@lspiro-Tick42 

I agree as well that many implementations are probably storing context this way.  The WG setting an imperative around implementation though is not needed here, so I don't see a justification for putting it in the standard.  

`getContextValues` is definitely something to consider.  Like `clearContext`, since it is additive to the current 1.1 proposal (and nothing is stoping implementors from prototyping these concepts today), I don't see any downside in considering these APIs post-1.1.

**Comment by mattjamieson** [comment]
@nkolba fair point that the meaning of "Last" is potentially unclear.

The reason for my suggestion to separate the features into two functions is that in implementations which are already taking the approach of keying by type (or in mine at least!) the concept of the untyped "current" context doesn't really make sense, but accessing the most recently received context is useful.

In the same vein, I'd argue that the equivalent to the untyped `addContextListener` is closer to @lspiro-Tick42's suggestion of returning `Array<Context>|null` (i.e. the untyped context listener receives all context, the untyped current context returns all context types). Though I think that's probably better implemented as a separate function as @lspiro-Tick42 suggested.

Possibly we could have `getCurrentContext(contextType: string)` and `getLatestContext()` (instead of _Last_) if the naming is the primary concern.

In any case, I am OK with the current API if no one else feels strongly; in which case I wonder if the documentation for `getCurrentContext` can be really explicit:

> Returns the most recent context that was broadcast on the channel. If no context has been set on the channel this will return `null`.
>
> If the optional `contextType` is provided, returns the current context if its type matches.
>
> If the desktop agent implementation has chosen to record contexts by type, supplying the optional `contextType` **may** return the most recently received context with a matching type.

**Comment by mattjamieson** [comment]
A different (possibly unrelated) thought...

If a client application has called `addContextListener("fdc3.contact", ...)` and an `fdc3.instrument` is broadcast, what do we expect to be the return of `getCurrentContext()`?

**Comment by lspiro-Tick42** [comment]
If we add comments making it clear what the expected behaviour is, then I am OK with deferring adding getContextValues() and clearContext() to version 1.2, we will add them to our FDC3 implementation, with comments making it clear that this is not in the FDC3 statndard.

I would not expect calls to addContextListener to change the behavious of getCurrentContext(), so I would expect getCurrentContext() to return the fdc3.instrument value whether or not the implementation support ‘per type’ context.

And for completeness, I would expect getCurrentContext(“fdc3.instrument”) to return the instrument irrespective of the implementation strategy, wheras I would expect getCurrentContext(“fdc3.contact”) to return the last contact value broadcast or null if no contact had been broadcast, for ‘per type’ Context implementations and to return null for a ‘single value’ Context implementation.

If we are talking about v 1.2 proposed changes, I would like to propose adding a
    broadcast( Array<Context> );
method to allow an application to set multiple versions of the same context, but I that would be for discussion after the v1.1 spec,

Leslie

**Comment by nkolba** [comment]
Thanks @mattjamieson and @lspiro-Tick42 !

I will make a PR taking a pass at making the current spec and API documentation more clear and explicit on these points.  We can then dig into multi-context, clearing channels, and other management questions post-1.1.

@mattjamieson - IMHO I don't think calls to `addContextListener` would impact the return value of `getCurrentContext`.

**Comment by mattjamieson** [comment]
Thanks both. @nkolba happy to collaborate on a PR if that's helpful.

And I agree on the `addContextListener`/`getCurrentContext` behaviour, I just wanted to check it wasn't just me! :grinning:

**Comment by nkolba** [comment]
Thanks @mattjamieson !   Will reach out on Monday

**Comment by nkolba** [comment]
@mattjamieson 

Here is my first pass at revised language:

When a _context type_ is provided, the most recent context matching the type will be returned, or _null_ if no matching context is found.

If no _context type_ is provided, the most recent context that was broadcast on the channel - regardless of type - will be returned.  If no context has been set on the channel, it will return _null_.

How recent contexts are stored is up to the specific Desktop Agent implementation.   For example, an implementation could store context history for a channel in a single array and search through the array for the last context matching a provided type, or context could be maintained as a dictionary keyed by context types.  

If getting the current context fails, the promise will return an `Error` with a string from the `ChannelError` enumeration.

**Comment by mattjamieson** [comment]
@nkolba 

Thanks. One small change: if we are not mandating that implementations support context history I think we should make that explicit.

> It is up to the specific Desktop Agent implementation whether and how recent contexts are stored. For example, an implementation could store context history for a channel in a single array and search through the array for the last context matching a provided type, or context could be maintained as a dictionary keyed by context types. An implementation could also choose not to support context history, in which case this method will return _null_ for any context type not matching the type of the most recent context.

**Comment by nkolba** [comment]
Thanks @mattjamieson - this looks good to me.  Will make a PR later today

**Comment by nkolba** [comment]
https://github.com/finos/FDC3/pull/182

**Comment by rikoe** [comment]
Closed by #182
